src/HttpFactory.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+declare(strict_types=1);
+
+namespace GuzzleHttp\Psr7;
+
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ServerRequestFactoryInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Psr\Http\Message\UploadedFileFactoryInterface;
+use Psr\Http\Message\UriFactoryInterface;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Message\StreamInterface;
+use Psr\Http\Message\UploadedFileInterface;
+use Psr\Http\Message\UriInterface;
+
+/**
+ * Implements all of the PSR-17 interfaces.
+ *
+ * Note: in consuming code it is recommended to require the implemented interfaces
+ * and inject the instance of this class multiple times.
+ */
+final class HttpFactory implements
+    RequestFactoryInterface,
+    ResponseFactoryInterface,
+    ServerRequestFactoryInterface,
+    StreamFactoryInterface,
+    UploadedFileFactoryInterface,
+    UriFactoryInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function createUploadedFile(
+        StreamInterface $stream,
+        int $size = null,
+        int $error = \UPLOAD_ERR_OK,
+        string $clientFilename = null,
+        string $clientMediaType = null
+    ): UploadedFileInterface {
+        if ($size === null) {
+            $size = $stream->getSize();
+        }
+
+        return new UploadedFile($stream, $size, $error, $clientFilename, $clientMediaType);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createStream(string $content = ''): StreamInterface
+    {
+        return \GuzzleHttp\Psr7\stream_for($content);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createStreamFromFile(string $file, string $mode = 'r'): StreamInterface
+    {
+        $resource = \GuzzleHttp\Psr7\try_fopen($file, $mode);
+
+        return \GuzzleHttp\Psr7\stream_for($resource);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createStreamFromResource($resource): StreamInterface
+    {
+        return \GuzzleHttp\Psr7\stream_for($resource);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createServerRequest(string $method, $uri, array $serverParams = []): ServerRequestInterface
+    {
+        if (empty($method)) {
```

Comment:
Shouldn't this be `$method === ''`?

I made the same comment to @Nyholm  [here ](https://github.com/guzzle/psr7/pull/210/files#r230211710) do you believe that this is also `premature optimization`... 

(first time that I heard this term by you @sagikazarmark... good to know 👍 )

Reply:
Honestly, I don't think it matters that much. You are free to create a PR if it bugs your OCD, but yeah, I would classify this as premature optimization 😄

Reply:
That logic seems to contradict what psr-17 says

> Note that server parameters are taken precisely as given - no parsing/processing
>  of the given values is performed. In particular, no attempt is made to
 >  determine the HTTP method or URI, which must be provided explicitly.